sds_data_manager/lambda_images/instruments/batch_starter.py

```diff
@@ -0,0 +1,425 @@
+    json_path = os.path.join(dir_path, "dependents.json")
+
+    with open(json_path) as f:
+        data = json.load(f)
+
+    instruments_to_process = []
+
+    for record in uningested:
+        dependencies = data[record["instrument"]][record["level"]]
+        result = query_dependents(cur, version, [record["date"]], dependencies)
+
+        # Check if all dependencies for this date are present in result
+        all_dependencies_met = all_dependency_present(result, dependencies)
+
+        if all_dependencies_met:
+            print(f"All dependencies for {record['date']} are met!")
+            instruments_to_process.append(
+                {
+                    "instrument": record["instrument"],
+                    "level": record["level"],
+                    "date": record["date"],
+                }
+            )
+        else:
+            print(f"Some dependencies for {record['date']} are missing!")
+
+    return instruments_to_process
+
+
+def all_dependency_present(result, dependencies):
+    """
+    Checks if all specified dependencies are present
+    in the given result.
+
+    Parameters
+    ----------
+    result : list of dict
+        Result of a query.
+    dependencies : list of dict
+        List of required dependencies.
+
+    Returns
+    -------
+    bool
+        True if all dependencies are found in
+        the `result`, otherwise False.
+
+    """
+    result_list = [{"instrument": r["instrument"], "level": r["level"]} for r in result]
+
+    # Convert dependencies to lowercase for comparison
+    dependencies = [
+        {"instrument": d["instrument"].lower(), "level": d["level"]}
+        for d in dependencies
+    ]
+
+    # Check if the items in dependencies are all present in result_list
+    if set(tuple(item.items()) for item in dependencies).issubset(
+        set(tuple(item.items()) for item in result_list)
+    ):
+        print("All dependencies are found.")
+        return True
+    else:
+        print("Some dependencies are missing.")
+        return False
+
+
+def prepare_data(instruments_to_process):
+    """
+    Groups input data by 'instrument' and 'level', and aggregates the dates
+    for each group into a list.
+
+    Parameters
+    ----------
+    instruments_to_process : list of dict
+        A list of dictionaries which is not aggregated.
+
+    Returns
+    -------
+    grouped_dict: dict
+        A dictionary of instruments, each containing a dictionary of
+        levels with a list of dates.
+    """
+    grouped_data = defaultdict(lambda: defaultdict(list))
+    for item in instruments_to_process:
+        instrument = item["instrument"]
+        level = item["level"]
+        date = item["date"]
+        grouped_data[instrument][level].append(date)
+
+    grouped_dict = {
+        instrument: dict(levels) for instrument, levels in grouped_data.items()
+    }
+
+    return grouped_dict
+
+
+def lambda_handler(event: dict, context):
```

Comment:
I think I'm confusing myself with the dependents vs dependencies. 

This is how I'm interpreting this function:
- An event is received because a new file was added to the S3
- That event contains the filename which is used to query for information about the file: level, version, etc.
- Another query is made to get the file's dependents
   - _This is the first place I'm getting confused. It seems like the dependents are all of the things that were needed to create this file, but it seems like it should be getting the dependencies which are all the files that this files is used to help create?_
- It checks to see if those dependents have been ingested and retains the ones that have not
- Another query is made to all of the uningested files' dependencies
   - _Same confusion as above, I thought dependencies were all the files that this file was used to help make, but it seems like it should be querying for all the files that are needed to create the uningested file which would be the dependents?_
 - Those files that come back from the previous query are input into the step function to be created
    - _What happens if somewhere down the dependency line it's missing something that it needs? Like the step function is trying to process an L1c file, but an L1a that it needs doesn't exist? It kind of seems like the dependency checks for uningested files should be recursive?_

I'm probably reading the json wrong and not totally understanding this function, but wanted to confirm.